apollo-router/src/plugins/record_replay/record.rs

```diff
@@ -0,0 +1,356 @@
+                    query: req.subgraph_request.body().query.clone(),
+                    operation_name: req.subgraph_request.body().operation_name.clone(),
+                    variables: req.subgraph_request.body().variables.clone(),
+                    headers: externalize_header_map(req.subgraph_request.headers())
+                        .expect("failed to externalize header map"),
+                },
+                move |req: RequestDetails, future| {
+                    let subgraph_name = subgraph_name.clone();
+                    async move {
+                        let res: subgraph::ServiceResult = future.await;
+
+                        let operation_name = req
+                            .operation_name
+                            .clone()
+                            .unwrap_or_else(|| "UnnamedOperation".to_string());
+
+                        let res = match res {
+                            Ok(res) => {
+                                let subgraph = Subgraph {
+                                    subgraph_name,
+                                    response: ResponseDetails {
+                                        headers: externalize_header_map(
+                                            &res.response.headers().clone(),
+                                        )
+                                        .expect("failed to externalize header map"),
+                                        chunks: vec![res.response.body().clone()],
+                                    },
+                                    request: req,
+                                };
+
+                                res.context
+                                    .upsert::<_, Subgraphs>(SUBGRAPHS, |mut value| {
+                                        value.insert(operation_name, subgraph);
+                                        value
+                                    })
+                                    .expect("failed to insert subgraph into context");
+                                Ok(res)
+                            }
+                            Err(err) => Err(err),
+                        };
+
+                        res
+                    }
+                },
+            )
+            .service(service)
+            .boxed()
+    }
+}
+
+async fn write_file(dir: Arc<Path>, path: &str, contents: &[u8]) -> Result<(), BoxError> {
+    let path = dir.join(path);
+    let dir = path.parent().unwrap();
+    fs::create_dir_all(dir).await?;
+    fs::write(path, contents).await?;
+    Ok(())
+}
+
+fn recording_enabled(context: &Context) -> bool {
+    context
+        .get::<_, bool>(RECORD)
+        .unwrap_or(None)
+        .unwrap_or(false)
+}
+
+fn is_introspection(query: String, schema: Arc<Schema>) -> bool {
+    let query = Query::parse(query, &schema, &Configuration::default()).expect("query must valid");
+    query.contains_introspection()
+}
+
+fn externalize_header_map(
```

Comment:
This looks like a copy of the co-processor `externalize_header_map()`. Maybe better to move the original `fn` into the src/services/external.rs source and have both plugins use that function?